go-controller/pkg/node/vrfmanager/vrf_manager.go

```diff
@@ -0,0 +1,321 @@
+		}
+	}
+	if vrf.enslaveInterface != "" {
+		err = enslaveInterfaceToVRF(vrf.name, vrf.enslaveInterface)
+		if err != nil {
+			return fmt.Errorf("failed to enslave interface %s into vrf device: %s, err: %w", vrf.enslaveInterface, vrf.name, err)
+		}
+	}
+	return nil
+}
+
+// AddVRF adds a VRF device into the node.
+func (vrfm *Controller) AddVRF(name, enslaveInterface string, table uint32) error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+
+	vrfDev, ok := vrfm.vrfs[name]
+	if ok {
+		klog.V(5).Infof("VRF Manager: vrf %s already found in the cache", name)
+		if vrfDev.enslaveInterface != enslaveInterface {
+			return fmt.Errorf("VRF Manager: enslave interface mismatch for vrf device %s", name)
+		}
+		if vrfDev.table != table {
+			return fmt.Errorf("VRF Manager: table id mismatch for vrf device %s", name)
+		}
+	} else {
+		vrfDev = vrf{name, table, enslaveInterface, false}
+		vrfm.vrfs[name] = vrfDev
+	}
+	return vrfm.sync(vrfDev)
+}
+
+// Repair deletes stale VRF device(s) on the host. This helps remove
+// device(s) for which DeleteVRF is never invoked.
+// Assumptions: 1) The validVRFs list must contain device for which AddVRF
+// is already invoked. 2) The device name(s) in validVRFs are suffixed
+// with -vrf.
+func (vrfm *Controller) Repair(validVRFs sets.Set[string]) error {
+	vrfm.mu.Lock()
+	defer vrfm.mu.Unlock()
+
+	links, err := util.GetNetLinkOps().LinkList()
+	if err != nil {
+		return fmt.Errorf("failed to list links on the node, err: %v", err)
+	}
+
+	for _, link := range links {
+		name := link.Attrs().Name
+		// Skip if the link is not a vrf type or name is not suffixed with -vrf.
+		if link.Type() != "vrf" || !strings.HasSuffix(name, types.VRFDeviceSuffix) {
+			continue
+		}
+		if !validVRFs.Has(name) {
+			err = util.GetNetLinkOps().LinkDelete(link)
+			if err != nil {
+				klog.Errorf("VRF Manager: error deleting stale vrf device %s, err: %v", name, err)
+			}
+		}
+		delete(vrfm.vrfs, name)
+	}
+	return nil
+}
+
+// DeleteVRF deletes given VRF device from the node.
+func (vrfm *Controller) DeleteVRF(name string) (err error) {
+	vrfm.mu.Lock()
+	defer func() {
+		if err == nil {
+			delete(vrfm.vrfs, name)
+		}
+		vrfm.mu.Unlock()
+	}()
+	vrf, ok := vrfm.vrfs[name]
+	if !ok {
+		klog.V(5).Infof("VRF Manager: vrf %s not found in cache for deletion", name)
+		return nil
+	}
+	vrf.delete = true
+	err = vrfm.deleteVRF(vrf)
+	if err != nil {
+		return fmt.Errorf("failed to delete vrf device %s, err: %w", vrf.name, err)
+	}
+	return nil
+}
+
+func (vrfm *Controller) deleteVRF(vrf vrf) error {
+	link, err := util.GetNetLinkOps().LinkByName(vrf.name)
+	if err != nil && util.GetNetLinkOps().IsLinkNotFoundError(err) {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to retrieve vrf device %s, err: %v", vrf.name, err)
+	}
+	return util.GetNetLinkOps().LinkDelete(link)
+}
+
+func getExistingEnslaveInterfaceForVRF(vrfLink netlink.Link) (string, error) {
```

Comment:
getSlaveInterfaceNamesForVRF(vrfLink netlink.Link) ([]string, error)

Reply:
maybe it makes more sense to return a list of links, rather than strings

Reply:
currently it only needs name of the slave interfaces,  so sticking with returning strings, otherwise i had to add extra logic at calling side to retrieve names from returned links.